feat/todo

danifromecuador opened the pull request:
## In this PR, I've:

- [x] Created a `feat/todo` branch
- [x] Created a `Todo` component that will allow the user to:
  - Add a new goal
  - Mark that goal as completed
  - Delete all completed goals
  - See the percentage of completed goals
- [x] Reused `Todo` component in 3 components: `Daily`, `Weekly`, and `Monthly` goals, those components are being called on the `Todos` component and the only one given prop is their corresponding slice of the store
- [x] All data for each component is saved in the global store that is managed by [Zustand](https://zustand-demo.pmnd.rs/)
- [x] You can still use the [Redux extension](https://chromewebstore.google.com/detail/redux-devtools/lmhkpmbekcpmknklioeibfkpmmfibljd) to see the global statue on the browser thanks to a Zustand feature even when this project doesn't use Redux
- [x] Saved goals components data on local storage, so that data will be preserved when refreshing the page
- [x] Conditional rendering for the `Delete All Completed` button and `Completed percentage indicator`
- [x] Applied minimal style and responsiveness, it will be improved in ahead PRs.
- [x] Behavior: 

https://github.com/user-attachments/assets/cf02d3aa-a6d8-4420-9d87-0a5def8df1ef

danifromecuador committed:
Create Todos component to be reused 4 times
branch feat/todo

danifromecuador committed:
Add minimal style to Todo & Todos components
branch feat/todo

danifromecuador committed:
Create todoDailySlice and Render todos array
branch feat/todo

danifromecuador committed:
Store all data for each Todo component in the Zustand store
branch feat/todo

danifromecuador committed:
Change names between Todos.jsx and Todo.jsx

danifromecuador committed:
Add an item to the todos or dones array in Store
branch feat/todo

danifromecuador committed:
Remove commented code
branch feat/todo

danifromecuador committed:
Remove completed Tasks component from Daily, Weekly, Monthly components group
branch feat/todo

danifromecuador committed:
Separate logic from store.js to improve readability
branch feat/todo

danifromecuador committed:
Remove an item from Daily todos when clicking on it
branch feat/todo

danifromecuador committed:
Push removed completed todo into dones array
branch feat/todo

danifromecuador committed:
Change the text color of completed or done items
branch feat/todo

danifromecuador committed:
Mark an item as todo in daily todo array
branch feat/todo

danifromecuador committed:
Sort todos and dones arrays by its id
branch feat/todo

danifromecuador committed:
Count completed or dones items to show completed percentage
branch feat/todo

danifromecuador committed:
Conditional rendering of completed span
branch feat/todo

danifromecuador committed:
Delete all completed goals
branch feat/todo

danifromecuador committed:
Shrin code by using && operator in ternary operator
branch feat/todo

danifromecuador committed:
Fix classname typo between Todos and Todo
branch feat/todo

danifromecuador committed:
Change cursor shape when overing an li tag
branch feat/todo

danifromecuador committed:
Conditional rendering of Delete All Completed Btn
depending if there is or not one or more completed or done goal
branch feat/todo

danifromecuador committed:
Improve style of Todo component
branch feat/todo

danifromecuador committed:
Fix Stylelint error
branch feat/todo

danifromecuador committed:
Install propTypes library and fix propType ESLint error
branch feat/todo

danifromecuador committed:
Save in localStorage todos and dones items for Daily Goals component
branch feat/todo

danifromecuador committed:
Create slices for weekly and monthly data on Store
branch feat/todo

danifromecuador committed:
Give an aditional prop to each function
to avoid creating add and other functions for each daily weekly and monthly
branch feat/todo

danifromecuador committed:
Render Daily Weekly and Monthly components with a new arg
branch feat/todo

danifromecuador committed:
Delete aditional prop from Todos components
because that prop will be added on the Store
branch feat/todo

danifromecuador committed:
Remove prop sliceID from Todo component
as this prop will be given in the Store
branch feat/todo

danifromecuador committed:
Replace the arg with a string when calling functions
branch feat/todo

danifromecuador committed:
Fix localStorage getting wrong data for dones array
branch feat/todo

danifromecuador committed:
Fix ESLint error
branch feat/todo

danifromecuador committed:
Improve layout on small screens
branch feat/todo

danifromecuador committed:
Fix ESLint errors
branch feat/todo

Diegogagan2587 reviewed (changes requested):
## @danifromecuador   Hi Buddy! , 
# Status: Almost there! :orange_circle: :orange_circle: :orange_circle: 

 :shipit: 
Great Job So far! 🎉 

### Highlights
- [x] Each section (Daily Goals, Weekly Goals and Monthly Goals) is adding or removing tasks.
- [x] Nice Job rendering the progress shown as percentage.
- [x] Timer plays a sound when time is over

### Suggested changes:
#### Specify action names
You may ask, what is the issue with that? Well, when the user executes an action like add a task, an action name will describe the action that has been executed, not adding action names will make debugging harder and the action would be named as `anonymous` in dev tools. as shown in the example below:
![image](https://github.com/user-attachments/assets/910f7cc6-0d39-4b20-b249-1e55f377304c)
The problem will occur if you want do debug, you will not know which action to check because they are named as `anonimous`. To prevent this problem specify an `action name` on each action that updates the state as shown in the example below:
``` js
// todo_logic.js
export const add = (set, input, sliceID) => set((state) => ({
  [sliceID]: {
    ...state[sliceID],
    todos: [...state[sliceID].todos, { id: Date.now(), content: input }]
  }
}), false, 'todo/add')
```
after adding an `action name` you should be able to debug properly if there is an error:
![image](https://github.com/user-attachments/assets/901de070-19c0-48f9-8da8-f185b680a18c)

### Summary of pending task

- [x] Specify action names for each action

## references
[Zustand Issue: Actions Names in Devtools](https://github.com/pmndrs/zustand/discussions/1668#:~:text=Traducir%20al%20espa%C3%B1ol-,You%20can%20specify%20action%20name%20in%20set.,-zustand/readme.md)
[Zustand Readme # Login Actions](https://github.com/pmndrs/zustand/blob/0b55a3d74b598aba2e41b09854e9237a4ce0f990/readme.md?plain=1#L425)

_If you think we should take another approach share it with the team!!! Thanks for your effort!_

Cheers, and Happy coding!👏👏👏

Please, remember to tag me in your question so I can receive the notification._

danifromecuador committed:
Log a specific action type for each `set` function
by passing a third parameter on Store functions when using Redux devtools
branch feat/todo

danifromecuador commented:
Thanks for the valuable feedback, dear @Diegogagan2587, The required changes were made and now all the actions have a name that will be helpful when debugging. 
![image](https://github.com/user-attachments/assets/447c205a-8890-4b5e-b470-3840261a515e)

Diegogagan2587 reviewed (approved):
## @danifromecuador   Hi Buddy! , 
# Status: Approved! :green_circle: :green_circle: :green_circle: 

 :shipit: 
Great Job So far! 🎉 

### Highlights
- [x] Each section (Daily Goals, Weekly Goals and Monthly Goals) is adding or removing tasks.
- [x] Nice Job rendering the progress shown as percentage.
- [x] Timer plays a sound when time is over
- [x]  Requested Changes were applied.

### Suggested changes:

- All suggested changes were applied.

_If you think we should take another approach share it with the team!!! Thanks for your effort!_

Cheers, and Happy coding!👏👏👏

Please, remember to tag me in your question so I can receive the notification._

danifromecuador commented:
Thanks, dear reviewer!
